twine/package.py

```diff
@@ -133,104 +133,51 @@ def from_filename(cls, filename: str, comment: Optional[str]) -> "PackageFile":
                 "Unknown distribution format: '%s'" % os.path.basename(filename)
             )
 
-        supported_metadata = list(pkginfo.distribution.HEADER_ATTRS)
-        if cls._is_unknown_metadata_version(captured):
+        # Parse and validate metadata.
+        meta, unparsed = metadata.parse_email(data)
+        if unparsed:
             raise exceptions.InvalidDistribution(
-                "Make sure the distribution is using a supported Metadata-Version: "
-                f"{', '.join(supported_metadata)}."
+                "Invalid distribution metadata: {}".format(
+                    ", ".join("unrecognized or malformed field {key!r}")
+                    for key in unparsed
+                )
             )
-        # If pkginfo <1.11 encounters a metadata version it doesn't support, it may give
-        # back empty metadata. At the very least, we should have a name and version,
-        # which could also be empty if, for example, a MANIFEST.in doesn't include
-        # setup.cfg.
-        missing_fields = [
-            f.capitalize() for f in ["name", "version"] if not getattr(meta, f)
-        ]
-        if missing_fields:
-            msg = f"Metadata is missing required fields: {', '.join(missing_fields)}."
-            if cls._pkginfo_before_1_11():
-                msg += (
-                    "\n"
-                    "Make sure the distribution includes the files where those fields "
-                    "are specified, and is using a supported Metadata-Version: "
-                    f"{', '.join(supported_metadata)}."
+        # setuptools emits License-File metadata fields while declaring
+        # Metadata-Version 2.1. This is invalid because the metadata
+        # specification does not allow to add arbitrary fields, and because
+        # the semantic implemented by setuptools is different than the one
+        # described in PEP 639. However, rejecting these packages would be
+        # too disruptive. Drop License-File metadata entries from the data
+        # sent to the package index if the declared metadata version is less
+        # than 2.4.
```

Comment:
Thanks for the explanation here. I'm worried this change could also be disruptive. What's the harm in allowing the metadata fields to pass as they had before? Is it necessary to support metadata 2.4 (this issue)?

Reply:
They are refused by PyPI.